### ISSUE TITLE
Dexcom One data not showing in practice measures

### ISSUE THREAD

**Issue by richiecroker** [ebmdatalab/openprescribing#3746]
@brianmackenna noted that Dexcom One sensors weren't appearing in the [Continuous Blood Glucose Monitoring](https://openprescribing.net/measure/cgm_sensors/national/england/) measure.

This measure had [recently been updated to use VMP and AMP codes](https://github.com/ebmdatalab/openprescribing/commit/abf832d1803e3c63e53ca8926bce83e426d4265f) to identify.  Other CGM devices are showing.

On investigation in BQ, the code used in the measure was shown to correctly identify Dexcom One:

```
SELECT bnf_code, nm FROM dmd.vmp WHERE id = 34865511000001109 -- selects bnf_codes from vmp table matching VMP id \n",
UNION DISTINCT -- joins vmp and amp tables together \n",
SELECT bnf_code, nm FROM dmd.amp WHERE vmp = 34865511000001109 AND bnf_code IS NOT NULL-- selects bnf_codes from amp table matching VMP id where they exist
```
<img width="254" alt="image" src="https://user-images.githubusercontent.com/18683239/200186922-c61edf87-8f42-4ea0-8191-94b49df4b530.png">

Further to that, code was run in BQ to find prescribing at a local practice, in order to mirror the measure data:

```
select bnf_name, p.bnf_code, SUM(items) as items from hscic.normalised_prescribing p
INNER JOIN
(SELECT bnf_code FROM dmd.vmp WHERE id = 34865511000001109 -- selects bnf_codes from vmp table matching VMP id \n",
    UNION DISTINCT -- joins vmp and amp tables together \n",
    SELECT bnf_code FROM dmd.amp WHERE vmp = 34865511000001109 AND bnf_code IS NOT NULL) bnf -- selects bnf_codes from amp table matching VMP id where they exist \n",
    ON p.bnf_code = bnf.bnf_code
WHERE
month between '2022-06-01' and '2022-08-01' and
p.bnf_code LIKE '2148000010%'
and practice = 'L83083'
group by bnf_name, bnf_code
```

Result:

![image](https://user-images.githubusercontent.com/18683239/200186968-7aa3a945-115e-4a89-8bd3-aa44107f6e61.png)

This compares with the measures page on OpenPrescribing, where Dexcom One is missing from the preparations breakdown.  

<img width="487" alt="image" src="https://user-images.githubusercontent.com/18683239/200187246-c6e11b8b-c670-4e3c-aa02-6c40c2c3fdb8.png">

Running the code below also shows that the data does not match the total quantities:

```
select month, SUM(quantity) as quantity, SUM(items) as items from hscic.normalised_prescribing p
INNER JOIN
(SELECT bnf_code FROM dmd.vmp WHERE id = 34865511000001109 -- selects bnf_codes from vmp table matching VMP id \n",
    UNION DISTINCT -- joins vmp and amp tables together \n",
    SELECT bnf_code FROM dmd.amp WHERE vmp = 34865511000001109 AND bnf_code IS NOT NULL) bnf -- selects bnf_codes from amp table matching VMP id where they exist \n",
    ON p.bnf_code = bnf.bnf_code
WHERE
month between '2022-06-01' and '2022-08-01' and
p.bnf_code LIKE '2148000010%'
and practice = 'L83083'
group by  month
order by month
```

Result:
<img width="209" alt="image" src="https://user-images.githubusercontent.com/18683239/200187080-45067c5d-3df4-4dfe-8dbf-2f10ba9e8df6.png">

compared with

<img width="254" alt="image" src="https://user-images.githubusercontent.com/18683239/200187168-a3e68610-1384-407d-b23a-d2de286a5bb5.png">

In summary, the website is showing different results to the same SQL ran in BQ

**Comment by richiecroker** [comment]
We are aware of the [VMP](https://openprescribing.net/dmd/vmp/34865511000001109/) and [AMP](https://openprescribing.net/dmd/amp/34865811000001107/) of the original Freestyle Libre having the same BNF code, The BSA have already responded to the error #3611

**Comment by richiecroker** [comment]
I've created a preview version of the measure (without changing any code), and this version seems correct:

England: 
![image](https://user-images.githubusercontent.com/18683239/200273750-6de82bc7-b2d2-4a73-b509-ad27df906588.png)

Local practice:

![image](https://user-images.githubusercontent.com/18683239/200273811-d5223833-5f8e-40e9-a049-a97985a04a1f.png)

It looks like a re-import would fix the issue, although i'm not happy to do this until a member of the team team has checked.

**Comment by inglesp** [comment]
`21480000103` (Dexcom One Sensor) is missing from this measure's list of numerator BNF codes:

```
>>> m = Measure.objects.get(pk="cgm_sensors")
>>> m.numerator_bnf_codes
['21480000100', '21480000101', '21480000102']
```

However, if we recompute this list now, it is present:

```
>>> from frontend.management.commands.import_measures import get_num_or_denom_bnf_codes
>>> get_num_or_denom_bnf_codes(m, "numerator", date(2022, 8, 1))
['21480000100', '21480000101', '21480000102', '21480000103']
```

Computing this list is done with this query:

```
        SELECT DISTINCT p.bnf_code
        FROM hscic.normalised_prescribing p -- joins prescribing data to subquery below, filtering list to BNF codes created in subquery 
 INNER JOIN -- joins the prescribing data table to the subquery thereby filtering the prescribing data to only these codes 
 (SELECT bnf_code FROM dmd.vmp WHERE id = 34865511000001109 -- selects bnf_codes from vmp table matching VMP id 
 UNION DISTINCT -- joins vmp and amp tables together 
 SELECT bnf_code FROM dmd.amp WHERE vmp = 34865511000001109 AND bnf_code IS NOT NULL) bnf -- selects bnf_codes from amp table matching VMP id where they exist 
 ON p.bnf_code = bnf.bnf_code
        WHERE 1 = 1
        ORDER BY bnf_code
```

I can't think of any reason why this would have returned the wrong thing.

However, this isn't the only measure affected:

```
>>> for m in Measure.objects.order_by("pk"):
...     if not m.numerator_is_list_of_bnf_codes:
...         continue
...     try:
...         numerator_bnf_codes = get_num_or_denom_bnf_codes(m, "numerator", date(2022, 8, 1))
...     except AttributeError:
...         print(f"Problem getting numerator BNF codes for {m.pk}")
...         continue
...     if numerator_bnf_codes != m.numerator_bnf_codes:
...         print(f"Mismatched numerator BNF codes for {m.pk}")
... 
Mismatched numerator BNF codes for aafpercent
Mismatched numerator BNF codes for bdzadq
Mismatched numerator BNF codes for bdzper1000
Mismatched numerator BNF codes for cgm_sensors
Mismatched numerator BNF codes for cmpaperpt
Mismatched numerator BNF codes for environmental_inhalers
Mismatched numerator BNF codes for fungal
Mismatched numerator BNF codes for glutenfree
Mismatched numerator BNF codes for injectable_antibiotics
Mismatched numerator BNF codes for ktt13_nsaids_ibuprofen
Mismatched numerator BNF codes for ktt9_antibiotics
Mismatched numerator BNF codes for lpfentanylir
Mismatched numerator BNF codes for lpglucosamine
Mismatched numerator BNF codes for lpherbal
Mismatched numerator BNF codes for lptravelvacs
Mismatched numerator BNF codes for lyrica
Mismatched numerator BNF codes for opioidome
Mismatched numerator BNF codes for preview_cost_mdi
Mismatched numerator BNF codes for preview_gabapentinoidsddd
Mismatched numerator BNF codes for preview_glutenfree
Mismatched numerator BNF codes for preview_pneumococcal
Mismatched numerator BNF codes for preview_refills
Mismatched numerator BNF codes for preview_salbutamol_per_1000
Mismatched numerator BNF codes for saba
Mismatched numerator BNF codes for sildenafil
Problem getting numerator BNF codes for silver
Mismatched numerator BNF codes for toothpaste
```

More digging required.

**Comment by inglesp** [comment]
There are two issues here:

#### Generating measures without the latest dm+d->BNF mapping

The numerator for the `cgm_sensors` measure is computed by finding the BNF codes of all the AMPs belonging to a [single VMP](https://openprescribing.net/dmd/vmp/34865511000001109/relationships/).

The Dexcom One AMP was added to the [BNF->dm+d mapping](https://www.nhsbsa.nhs.uk/prescription-data/understanding-our-data/bnf-snomed-mapping) in the August 2022 spreadsheet, which was published on 17th October, but we did not import the spreadsheet until 24th October as part of the weekly dm+d import.

However, we imported the August prescribing data (and generated the measures) on 19th-20th October.  So when the prescribing data was imported (and the measures generated), we did not know about the BNF code for the Dexcom One AMP.

This may have happened to other before without our noticing.  We can mitigate against this by not importing the prescribing data until the dm+d data has been imported.

#### BNF code changes

All other differences can be explained by us having run `generate_presentation_replacements` recently (#3773).  

## Next steps

Reimporting measures will rectify the problem.
